packages/monitor-v2/src/monitor-og/oSnapAutomation.ts

```diff
@@ -0,0 +1,393 @@
+  return allProposals.filter((proposal) => !deletedAssertionIds.includes(proposal.args.assertionId));
+};
+
+// Checks if a safeSnap safe from Snapshot proposal is supported by oSnap automation.
+const isSafeSupported = (safe: SafeSnapSafe, supportedModules: SupportedModules, chainId: number): boolean => {
+  for (const ogAddress in supportedModules) {
+    if (isMatchingSafe(safe, chainId, ogAddress)) return true;
+  }
+  return false;
+};
+
+// Filters out all Snapshot proposals that have been proposed on-chain. This is done by matching safe, explanation and
+// proposed transactions.
+const filterPotentialProposals = (
+  supportedProposals: SnapshotProposalExpanded[],
+  onChainProposals: TransactionsProposedEvent[],
+  params: MonitoringParams
+): SnapshotProposalExpanded[] => {
+  return supportedProposals.filter((supportedProposal) => {
+    const matchingOnChainProposals = onChainProposals.filter((onChainProposal) => {
+      // Check if safe and explanation match
+      if (
+        isMatchingSafe(supportedProposal.safe, params.chainId, onChainProposal.address) &&
+        supportedProposal.ipfs === tryHexToUtf8String(onChainProposal.args.explanation)
+      ) {
+        // Check if proposed transactions match
+        return onChainTxsMatchSnapshot(onChainProposal, supportedProposal.safe);
+      }
+      return false;
+    });
+    // Exclude Snapshot proposals with matching on-chain proposals
+    return matchingOnChainProposals.length === 0;
+  });
+};
+
+// Verifies proposals before they are proposed on-chain.
+const filterVerifiedProposals = async (
+  proposals: SnapshotProposalExpanded[],
+  supportedModules: SupportedModules,
+  params: MonitoringParams
+): Promise<SnapshotProposalExpanded[]> => {
+  // Convert expanded proposals back to GraphqlData format as this is used in SnapshotVerification.
+  const graphqlData: GraphqlData = {
+    proposals: proposals.map((proposal) => {
+      const { safe, ...clonedObject } = proposal;
+      return { ...clonedObject, plugins: { safeSnap: { safes: [safe] } } };
+    }),
+  };
+
+  // Verify all potential proposals.
+  const lastTimestamp = await getBlockTimestamp(params.provider, params.blockRange.end);
+  const verifiedProposals = (
+    await Promise.all(
+      graphqlData.proposals.map(async (proposal) => {
+        // Check that the proposal was approved properly on Snapshot assuming we are at the end block timestamp.
+        const voteOutcomVerified = verifyVoteOutcome(proposal, lastTimestamp, 0).verified;
+
+        // Check that proposal is hosted on IPFS and its content matches.
+        const ipfsVerified = (await verifyIpfs(proposal, params)).verified;
+
+        // Check that the proposal meets rules requirements for the target oSnap module.
+        const rulesVerified = verifyRules(
+          getModuleParameters(proposal.plugins.safeSnap.safes[0].umaAddress, supportedModules).parsedRules,
```

Comment:
Is it possible to have multiple safes in this array? Could a user induce this behavior to confuse the bot into proposing something that's invalid?

Reply:
No, this array will always have 1 safe. Though its possible for space admin to add multiple safes, we expand the Snapshot proposal to include only 1 safe [here](https://github.com/UMAprotocol/protocol/blob/406f84126861d8d5da98c9267ead72f674475072/packages/monitor-v2/src/monitor-og/oSnapAutomation.ts#L152-#L156), but in order to reuse the verification logic we convert this back to `plugins.safeSnap.safes` with 1 array member [here](https://github.com/UMAprotocol/protocol/blob/406f84126861d8d5da98c9267ead72f674475072/packages/monitor-v2/src/monitor-og/oSnapAutomation.ts#L238-#L244).